new-log-viewer/src/contexts/StateContextProvider.tsx

```diff
@@ -68,65 +73,107 @@ const STATE_DEFAULT: Readonly<StateContextType> = Object.freeze({
  *
- * @param beginLineNumToLogEventNum
- * @return The last log event number.
+ * @param navAction Action to navigate to a new page.
+ * @param currentPageNum
+ * @param numPages
+ * @return `PAGE_NUM` cursor.
  */
-const getLastLogEventNum = (beginLineNumToLogEventNum: BeginLineNumToLogEventNumMap) => {
-    const allLogEventNums = Array.from(beginLineNumToLogEventNum.values());
-    let lastLogEventNum = allLogEventNums.at(-1);
-    if ("undefined" === typeof lastLogEventNum) {
-        lastLogEventNum = 1;
+const getPageNumCursor = (
+    navAction: NavigationAction,
+    currentPageNum: number,
+    numPages: number
+): Nullable<CursorType> => {
+    if (STATE_DEFAULT.pageNum === currentPageNum) {
+        // eslint-disable-next-line no-warning-comments
+        // TODO: This shouldn't be possible, but currently, the page nav buttons remain enabled
+        // even when a file hasn't been loaded.
+        console.error("Page actions cannot be executed if the current page is not set.");
+
+        return null;
     }
 
-    return lastLogEventNum;
+    let newPageNum: number;
+    let position: EVENT_POSITION_ON_PAGE;
+    switch (navAction.code) {
+        case ACTION_NAME.SPECIFIC_PAGE:
+            position = EVENT_POSITION_ON_PAGE.TOP;
+
+            // Clamp is to prevent someone from requesting non-existent page.
+            newPageNum = clamp(navAction.args.pageNum, 1, numPages);
+            break;
+        case ACTION_NAME.FIRST_PAGE:
+            position = EVENT_POSITION_ON_PAGE.TOP;
+            newPageNum = 1;
+            break;
+        case ACTION_NAME.PREV_PAGE:
+            position = EVENT_POSITION_ON_PAGE.BOTTOM;
+            newPageNum = clamp(currentPageNum - 1, 1, numPages);
+            break;
+        case ACTION_NAME.NEXT_PAGE:
+            position = EVENT_POSITION_ON_PAGE.TOP;
+            newPageNum = clamp(currentPageNum + 1, 1, numPages);
+            break;
+        case ACTION_NAME.LAST_PAGE:
+            position = EVENT_POSITION_ON_PAGE.BOTTOM;
+            newPageNum = numPages;
+            break;
+        default:
+            return null;
+    }
+
+    return {
+        code: CURSOR_CODE.PAGE_NUM,
+        args: {pageNum: newPageNum, eventPositionOnPage: position},
+    };
 };
 
 /**
- * Sends a post message to a worker with the given code and arguments. This wrapper around
- * `worker.postMessage()` ensures type safety for both the request code and its corresponding
- * arguments.
+ * Submits a `LOAD_PAGE` request to a worker.
  *
  * @param worker
- * @param code
- * @param args
+ * @param cursor
  */
-const workerPostReq = <T extends WORKER_REQ_CODE>(
-    worker: Worker,
-    code: T,
-    args: WorkerReq<T>
+const loadPageByCursor = (
+    worker: Nullable<Worker>,
+    cursor: CursorType,
 ) => {
-    worker.postMessage({code, args});
+    if (null === worker) {
+        console.error("Unexpected null worker");
+
+        return;
+    }
```

Comment:
This probably seems extra, but I think it should be the caller's responsibility to check for worker being invalid. As far as I can tell, the two call sites where this is called don't really make sense unless worker is valid, so I think it would be good to do this check near the top of the caller functions.